### PR TITLE
feat: Support for embedding/retriever model distillation

### PR DESCRIPTION
**Pull request by vinay-raman**
# What does this PR do ?
This PR enables support for knowledge distillation of embedding/retriever models

# Changelog
feat: 
- Support for knowledge distillation (KD) of embedding/retriever models
- Loss functions for KD supported: Cosine loss, MSE loss, InfoNCE loss and combinations thereof
- Embedding dimension mismatch handled by trainable linear projection of student model onto teacher's embedding space 

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: examples/retrieval/distillation/ministral3_2b_distill.yaml [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  padding_side: left
+  force_hf: true
+
+dataloader:
+  _target_: torchdata.stateful_dataloader.StatefulDataLoader
+  dataset:
+    _target_: nemo_automodel.components.datasets.llm.make_retrieval_dataset
+    model_type: bi_encoder
+    data_dir_list:
+      - /lustre/fsw/portfolios/datascience/projects/datascience_rapids_research/retriever_models_research/01_datasets/03_training_datasets/nqsh_shuffled_50k.json
```
Comment (rnyak) [inline examples/retrieval/distillation/ministral3_2b_distill.yaml]: @vinay-raman we cannot use our lustre path here. please change it as in this example : https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/retrieval/bi_encoder/llama3_2_1b.yaml#L56-L58

File: examples/retrieval/distillation/ministral3_2b_distill.yaml [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  global_batch_size: 64
+  local_batch_size: 4
+  ckpt_every_steps: 500
+  num_epochs: 1
+
+model:
+  _target_: nemo_automodel.components.models.biencoder.student_with_projection.StudentWithProjection.build
+  pretrained_model_name_or_path: /lustre/fsw/portfolios/llmservice/users/viraman/ministral3/2B_pruned_model_rank
+  teacher_hidden_size: 3072
+  pooling: last
```
Comment (rnyak) [inline examples/retrieval/distillation/ministral3_2b_distill.yaml]: this should be "avg".

File: nemo_automodel/_transformers/retrieval.py [hunk 3]
```diff
@@ -152,6 +152,8 @@ def pool(last_hidden_states: torch.Tensor, attention_mask: torch.Tensor, pool_ty
         Pooled embeddings [batch_size, hidden_size]
     """
     last_hidden = last_hidden_states.masked_fill(~attention_mask[..., None].bool(), 0.0)
+    if pool_type == "mean":
+        pool_type = "avg"
```
Comment (rnyak) [inline nemo_automodel/_transformers/retrieval.py:158]: @vinay-raman do we need this condition? can we not set the pool_type "avg" automatically, in yaml files like here: https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/retrieval/bi_encoder/llama3_2_1b.yaml#L39

is that bcs the distillation source code uses "mean" instead?

File: nemo_automodel/components/models/biencoder/nemotron_embedding_model.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    
+    DEFAULT_MODEL_PATH = "/lustre/fsw/portfolios/llmservice/users/viraman/action_selection_evals/models/4b_nanov3_sft"
+    
+    def __init__(
+        self,
+        model_path: str = None,
+        device: str = None,
+        dtype: torch.dtype = torch.float16,
+        normalize_embeddings: bool = True,
+        pooling_strategy: str = "eos",  # "eos", "mean", "last"
```
Comment (rnyak) [inline nemo_automodel/components/models/biencoder/nemotron_embedding_model.py]: why eos?

File: examples/retrieval/distillation/ministral3_2b_distill.yaml [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+step_scheduler:
+  global_batch_size: 64
+  local_batch_size: 4
+  ckpt_every_steps: 500
+  num_epochs: 1
+
+model:
+  _target_: nemo_automodel.components.models.biencoder.student_with_projection.StudentWithProjection.build
+  pretrained_model_name_or_path: /lustre/fsw/portfolios/llmservice/users/viraman/ministral3/2B_pruned_model_rank
```
Comment (rnyak) [inline examples/retrieval/distillation/ministral3_2b_distill.yaml]: Please remove this path. only add a placeholder like "your_student_model_path"

File: examples/retrieval/distillation/ministral3_2b_distill.yaml [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  pooling: last
+  l2_normalize: false
+  capture_layers: []
+  trust_remote_code: true
+  torch_dtype: bfloat16
+  attn_implementation: flash_attention_2
+
+teacher_model:
+  _target_: nemo_automodel.components.models.biencoder.student_with_projection.TeacherEmbeddingEncoder.build
+  pretrained_model_name_or_path: /lustre/fsw/portfolios/llmservice/users/ybabakhin/model_checkpoints/best_models/ministral3_3B_bidirectional_instruct_release_lr_2e-6_last_pooling_03-04-150645/LATEST/model/consolidated
```
Comment (rnyak) [inline examples/retrieval/distillation/ministral3_2b_distill.yaml]: please remove this path. only add place holder, like "your_teacher_model_path"

File: nemo_automodel/recipes/retrieval/distill_bi_encoder.py [hunk 7]
```diff
@@ -0,0 +1,632 @@
...
+
+        self.teacher_model = None
+        if self.use_cached_teacher:
+            logger.info(
+                "Using cached teacher embeddings from %s; live teacher model will not be built.",
+                self.teacher_embeddings_cache,
+            )
+        else:
+            with torch.random.fork_rng(enabled=False):
+                self.teacher_model = self.cfg.teacher_model.instantiate(
```
Comment (yuhezhang-ai) [inline nemo_automodel/recipes/retrieval/distill_bi_encoder.py:232]: This live-teacher path appears to pass obsolete separate distributed kwargs into `teacher_model.instantiate()`. `TeacherEmbeddingEncoder.build()` forwards `**kwargs` to `NeMoAutoModelBiEncoder.from_pretrained()`, and current AutoModel rejects separate distributed kwargs such as `moe_mesh` / `distributed_config`; it expects `distributed_setup`. The default examples without `teacher_embeddings_cache` will hit this path and fail during setup. Can this pass `distributed_setup=self.distributed_setup` instead and avoid forwarding the separate distributed kwargs?

File: nemo_automodel/components/models/biencoder/nemotron_bidirectional_embedding.py [hunk 8]
```diff
@@ -0,0 +1,538 @@
...
+)
+
+try:
+    from nemo_automodel.components.models.biencoder.state_dict_adapter import BiencoderStateDictAdapter
+except ImportError:
+    BiencoderStateDictAdapter = object
+
+from nemo_automodel.shared.import_utils import get_check_model_inputs_decorator
+
+logger = logging.get_logger(__name__)
```
Comment (yuhezhang-ai) [inline nemo_automodel/components/models/biencoder/nemotron_bidirectional_embedding.py]: This module imports Python stdlib `logging`, which exposes `getLogger()`, not `get_logger()`. The added test imports this module at collection time, so `pytest tests/unit_tests/models/biencoder/test_nemotron_embedding.py` currently errors before running tests.

File: nemo_automodel/recipes/retrieval/distill_bi_encoder.py [hunk 9]
```diff
@@ -0,0 +1,632 @@
...
+        dp_size = self._get_dp_group_size(include_cp=True)
+        if dp_size <= 1:
+            return
+        for param in params:
+            if param.grad is None:
+                continue
+            torch.distributed.all_reduce(param.grad, op=torch.distributed.ReduceOp.SUM, group=dp_group)
+            param.grad.div_(dp_size)
+
+    def _build_optimizer_param_groups(self) -> list[dict[str, Any]]:
```
Comment (yuhezhang-ai) [inline nemo_automodel/recipes/retrieval/distill_bi_encoder.py:334]: This override is not currently used: `TrainBiEncoderRecipe.setup()` still builds optimizer param groups inline instead of calling `self._build_optimizer_param_groups()`. As a result the distillation projection group is ignored, so `projection_lr` and the no-weight-decay treatment for `projection.weight` do not apply. The base setup should use this hook before optimizer construction.

**Review by yuhezhang-ai — COMMENTED** [review]
A few issues I noticed in the distillation path.

File: nemo_automodel/components/models/biencoder/nemotron_bidirectional_embedding.py (hunk 8 above)
Comment (vinay-raman) [reply nemo_automodel/components/models/biencoder/nemotron_bidirectional_embedding.py]: This file has been removed

[the remark this answers is not in the document]
Comment (vinay-raman) [reply nemo_automodel/components/models/biencoder/nemotron_bidirectional_embedding.py]: This file is not needed, has been removed.

File: examples/retrieval/distillation/ministral3_2b_distill.yaml (hunk 1 above)
Comment (vinay-raman) [reply examples/retrieval/distillation/ministral3_2b_distill.yaml]: fixed

File: examples/retrieval/distillation/ministral3_2b_distill.yaml (hunk 2 above)
Comment (vinay-raman) [reply examples/retrieval/distillation/ministral3_2b_distill.yaml]: fixed

File: nemo_automodel/_transformers/retrieval.py (hunk 3 above)
Comment (vinay-raman) [reply nemo_automodel/_transformers/retrieval.py:158]: fixed everything to 'avg'

File: nemo_automodel/components/models/biencoder/nemotron_embedding_model.py (hunk 4 above)
Comment (vinay-raman) [reply nemo_automodel/components/models/biencoder/nemotron_embedding_model.py]: this file is not needed, removed the file

File: examples/retrieval/distillation/ministral3_2b_distill.yaml (hunk 5 above)
Comment (vinay-raman) [reply examples/retrieval/distillation/ministral3_2b_distill.yaml]: fixed

File: examples/retrieval/distillation/ministral3_2b_distill.yaml (hunk 6 above)
Comment (vinay-raman) [reply examples/retrieval/distillation/ministral3_2b_distill.yaml]: fixed

File: nemo_automodel/recipes/retrieval/distill_bi_encoder.py (hunk 7 above)
Comment (vinay-raman) [reply nemo_automodel/recipes/retrieval/distill_bi_encoder.py:232]: fixed

File: nemo_automodel/recipes/retrieval/distill_bi_encoder.py (hunk 9 above)
Comment (vinay-raman) [reply nemo_automodel/recipes/retrieval/distill_bi_encoder.py:334]: fixed

**Comment by rnyak** [comment]
@vinay-raman let's close this PR.
